pkg/kv/kvprober/kvprober.go

```diff
@@ -257,9 +257,19 @@ func (p *Prober) Start(ctx context.Context, stopper *stop.Stopper) error {
 					return
 				}
 
-				probeCtx, sp := tracing.EnsureChildSpan(ctx, p.tracer, opName+" - probe")
-				probe(probeCtx, p.db, pl)
-				sp.Finish()
+				func() {
+					// KVPROBER_TRACES is for traces of probes sent to the KV layer by
+					// kvprober. Note that lower fidelity logging goes to the HEALTH
+					// channel instead of the KVPROBER_TRACES channel. It is better
+					// to store traces using proper logging infra, but such infra
+					// is not always available, hence this stopgap.
+					probeCtx, collectAndFinish := tracing.ContextWithRecordingSpan(
```

Comment:
FWIW, one way I imagined this stuff could work is to have a `heap_profiler` for such traces. That is, instead of (or in addition to) dumping walls of text into the logs, we have a [Dumper](https://github.com/cockroachdb/cockroach/blob/a74d71d895980281d7dabec26711b274811d8727/pkg/server/dumpstore/dumpstore.go#L51) somewhere that takes kvprober traces and writes them out in [Jaeger format](https://github.com/cockroachdb/cockroach/blob/775c2b0b0302ebf4b0c8c8515e015b793b8585ce/pkg/util/tracing/recording.go#L322-L328), where they can easily be explored (using [jaeger](https://www.jaegertracing.io/)).

I think this use case inverts what the Dumper was made for somewhat - the dumper does action X every Y seconds and grooms the result set - here it gets handed something at arbitrary points in time and grooms - but I think it could be accommodated.

This might be too big a lift for @joshimhoff but perhaps we can discuss the idea here. The proposed mechanism for trace collection would also be helpful for #33007 and I am sure we would find a few more places where it ends up helpful; for example once we are at the place where we can start tracing requests that are "slow", the trace has to be collected somewhere as well.

Reply:
I like this idea a lot!

> I think this use case inverts what the Dumper was made for somewhat - the dumper does action X every Y seconds and grooms the result set - here it gets handed something at arbitrary points in time and grooms - but I think it could be accommodated.

Yes we can alter the design to make the retention policy more flexible.